File: 31_oct/problem_08_evacuate_to_moon/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    long long maxEnergy(int n, int m, int h, vector<int>& capacity, vector<int>& power) {
        // Sort both arrays in descending order to pair largest with largest
        sort(capacity.begin(), capacity.end(), greater<int>());
        sort(power.begin(), power.end(), greater<int>());
        
        long long totalEnergy = 0;
        int pairs = min(n, m);
        
        for (int i = 0; i < pairs; i++) {
            // Energy that can be generated in h hours
            long long generatedEnergy = (long long)power[i] * h;
            // Take minimum of generated and capacity
            totalEnergy += min(generatedEnergy, (long long)capacity[i]);
        }
        
        return totalEnergy;
    }
};

int main() {
    int t;
    cin >> t;
    
    while (t--) {
        int n, m, h;
        cin >> n >> m >> h;
        
        vector<int> capacity(n), power(m);
        for (int i = 0; i < n; i++) {
            cin >> capacity[i];
        }
        for (int i = 0; i < m; i++) {
            cin >> power[i];
        }
        
        Solution solution;
        long long result = solution.maxEnergy(n, m, h, capacity, power);
        cout << result << endl;
    }
    
    return 0;
}

File: 31_oct/problem_08_evacuate_to_moon/solution.py
class Solution(object):
    def maxEnergy(self, n, m, h, capacity, power):
        """
        :type n: int
        :type m: int
        :type h: int
        :type capacity: List[int]
        :type power: List[int]
        :rtype: int
        """
        # Sort both arrays in descending order
        capacity.sort(reverse=True)
        power.sort(reverse=True)
        
        total_energy = 0
        pairs = min(n, m)
        
        for i in range(pairs):
            # Energy that can be generated in h hours
            generated_energy = power[i] * h
            # Take minimum of generated and capacity
            total_energy += min(generated_energy, capacity[i])
        
        return total_energy

def solve(infile, outfile):
    """Standalone execution for orchestrator"""
    t = int(infile.readline())
    results = []
    
    for _ in range(t):
        n, m, h = map(int, infile.readline().split())
        capacity = list(map(int, infile.readline().split()))
        power = list(map(int, infile.readline().split()))
        
        solution = Solution()
        result = solution.maxEnergy(n, m, h, capacity, power)
        results.append(result)
    
    for r in results:
        outfile.write(str(r) + '\n')

if __name__ == "__main__":
    import sys
    solve(sys.stdin, sys.stdout)

